File: empresa-back/src/main/java/SoulCode/Empresa/services/SupervisorService.java
package SoulCode.Empresa.services;

import java.util.List;
import java.util.Optional;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import SoulCode.Empresa.models.Cargo;
import SoulCode.Empresa.models.Funcionario;
import SoulCode.Empresa.models.Supervisor;
import SoulCode.Empresa.repositorys.CargoRepository;
import SoulCode.Empresa.repositorys.SupervisorRepository;



@Service
public class SupervisorService {

	@Autowired
	private SupervisorRepository supervisorRepository;
	@Autowired
	private CargoService cargoService;
	@Autowired
	private CargoRepository cargoRepository;
	
	
	
// GET todos Supervisores
	public List<Supervisor> mostrarTodosSupervisores(){
		List<Supervisor> supervisor = supervisorRepository.findAll();
		return supervisor;
	}
	
//	GET único Supervisor
	public Supervisor mostrarUmSupervisor(Integer id_supervisor) {
		Optional<Supervisor> supervisor = supervisorRepository.findById(id_supervisor);
		return supervisor.orElseThrow();
	}
	
//	Buscar supervisor pelo nome
	public Supervisor buscarSupervisorPeloNome(String sup_nome){
		Supervisor supervisor = supervisorRepository.fetchByName(sup_nome);
		return supervisor;
	}
	
	
//	Supervisor sem cargo
	public List<Supervisor> supervisorSemCargo(){
		return supervisorRepository.supervisorSemCargo();
	}
	
	
//	Cadastrar Supervisor
	public Supervisor cadastrarSupervisor(Integer id_cargo , Supervisor supervisor) {
		supervisor.setId_supervisor(null);
		if ( id_cargo != null) {
			Cargo cargo = cargoService.buscarUmCargo(id_cargo);
			supervisor.setCargo(cargo);
			cargo.setSupervisor(supervisor);
		}
		return supervisorRepository.save(supervisor);
		
	}
	
//	Editar Supervisor
	public Supervisor editarSupervisor(Supervisor supervisor) {
		mostrarUmSupervisor(supervisor.getId_supervisor());
		return supervisorRepository.save(supervisor);
	}
	
//	GET de supervisor que está dentro de um cargo
	
	public Supervisor buscaSupervisorDoCargo(Integer id_cargo) {
		Supervisor supervisor = supervisorRepository.buscaSupervisorDoCargo(id_cargo);
		return supervisor;
	}
	
//	Salvar foto
	public Supervisor salvarFoto(Integer id_supervisor, String caminhoFoto) {
		Supervisor supervisor = mostrarUmSupervisor(id_supervisor);
		supervisor.setSup_foto(caminhoFoto);
		return supervisorRepository.save(supervisor);
	}
	

//	DELETE Supervisor
	public void deletarUmSupervisor(Integer id_supervisor) {
		supervisorRepository.deleteById(id_supervisor);
	}
	
	
//	Buscar supervisores com seu cargo
	public List<List> supervisorComSeuCargo(){
		return supervisorRepository.supervisorComSeuCargo();
	}
	
//	Retirar supervisor de um cargo
	public Supervisor deixarSupervisorSemCargo(Integer id_supervisor) {
		Supervisor supervisor = mostrarUmSupervisor(id_supervisor);
		supervisor.setCargo(null);
		return supervisorRepository.save(supervisor);
	}


}
	
	
	

File: empresa-front/empresaFront/src/app/services/supervisor.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Cargo } from '../models/cargo';
import { Supervisor } from '../models/supervisor';

@Injectable({
  providedIn: 'root'
})
export class SupervisorService {

  baseUrl: String = 'http://localhost:8080/empresa'

  constructor(private http:HttpClient) { }

  buscaSupervisorDoCargo(id_cargo:String):Observable<Supervisor>{

    const url = `${this.baseUrl}/supervisor-cargo/${id_cargo}`
    return this.http.get<Supervisor>(url)

  }

  mostrarUmSupervisor(id_supervisor:String):Observable<Supervisor>{
    const url = `${this.baseUrl}/supervisor/${id_supervisor}`
    return this.http.get<Supervisor>(url)
  }

  supervisorComSeuCargo():Observable<any[]>{
    const url = `${this.baseUrl}/supervisor-cargo`
    return this.http.get<any[]>(url)
  }

  buscarSupervisorSemCargo():Observable<Supervisor[]>{

    const url = `${this.baseUrl}/supervisorSemCargo`
    return this.http.get<Supervisor[]>(url)

  }

  mostrarTodosSupervisores():Observable<Supervisor[]>{
    const url = `${this.baseUrl}/supervisor`
    return this.http.get<Supervisor[]>(url)
  }

    cadastrarSupervisor(supervisor:Supervisor):Observable<Supervisor>{
    const url = `${this.baseUrl}/supervisor`
    return this.http.post<Supervisor>(url,supervisor)

  }

  buscarSupervisorPeloNome(sup_nome:String):Observable<Supervisor>{
    const url = `${this.baseUrl}/supervisor-nome/${sup_nome}`
    return this.http.get<Supervisor>(url)
  }

  deixarSupervisorSemCargo(supervisor:Supervisor , id_supervisor:String ):Observable<Supervisor>{
    const url = `${this.baseUrl}/supervisorSemCargo/${id_supervisor}`
    return this.http.put<Supervisor>(url,supervisor)


  }

  editarSupervisor(id_cargo:any , id_supervisor:any , supervisor:Supervisor):Observable<Supervisor>{
    const url = `${this.baseUrl}/supervisor/${id_supervisor}?cargo=${id_cargo}`
    return this.http.put<Supervisor>(url,supervisor)
  }

  editarSomenteSupervisor(id_supervisor:any , supervisor:Supervisor):Observable<Supervisor>{
    const url = `${this.baseUrl}/editarSomenteSupervisor/${id_supervisor}`
    return this.http.put<Supervisor>(url,supervisor)
  }

  excluirSupervisor(id_supervisor:any):Observable<void>{
    const url = `${this.baseUrl}/supervisor/${id_supervisor}`
    return this.http.delete<void>(url)
  }

}
